daemons/ipa-kdb/ipa_kdb_mspac.c

```diff
@@ -1835,27 +1840,60 @@ krb5_error_code filter_logon_info(krb5_context context,
     /* check netbios/flat name */
     if (strcasecmp(info->info->info3.base.logon_domain.string,
                    domain->flat_name) != 0) {
-        krb5_klog_syslog(LOG_ERR, "PAC Info mismatch: domain = %s, "
-                                  "expected flat name = %s, "
-                                  "found logon name = %s",
-                                  domain->domain_name, domain->flat_name,
-                                  info->info->info3.base.logon_domain.string);
-        return EINVAL;
+        struct ipadb_mspac *mspac_ctx = ipactx->mspac;
+        result = FALSE;
+        /* Didn't match but perhaps the original PAC was issued by a child domain's DC? */
+        if ((mspac_ctx != NULL) && (domain->parent == NULL)) {
+            /* this is forest root, search for any domains under it */
+            for (k = 0; k < mspac_ctx->num_trusts; k++) {
+                if ((mspac_ctx->trusts[k].parent == domain) &&
+                    strcasecmp(info->info->info3.base.logon_domain.string,
+                               mspac_ctx->trusts[k].flat_name) == 0) {
+                    result = TRUE;
+                    break;
+                }
+            }
+        }
+        if (result == FALSE) {
+            krb5_klog_syslog(LOG_ERR, "PAC Info mismatch: domain = %s, "
+                                    "expected flat name = %s, "
+                                    "found logon name = %s",
+                                    domain->domain_name, domain->flat_name,
+                                    info->info->info3.base.logon_domain.string);
+            return EINVAL;
+        }
     }
 
     /* check exact sid */
     result = dom_sid_check(&domain->domsid, info->info->info3.base.domain_sid, true);
     if (!result) {
-        domstr = dom_sid_string(NULL, info->info->info3.base.domain_sid);
-        if (!domstr) {
+        struct ipadb_mspac *mspac_ctx = ipactx->mspac;
+        result = FALSE;
+        /* Didn't match but perhaps the original PAC was issued by a child domain's DC? */
+        if ((mspac_ctx != NULL) && (domain->parent == NULL)) {
+            /* this is forest root, search for any domains under it */
+            for (k = 0; k < mspac_ctx->num_trusts; k++) {
+                if (mspac_ctx->trusts[k].parent == domain) {
+                    result = dom_sid_check(&mspac_ctx->trusts[k].domsid,
+                                     info->info->info3.base.domain_sid, true);
+                }
+                if (result) {
+                    break;
+                }
```

Comment:
Similar to above, I don't know how useful it would be to know the non-matching SIDs in case of failure.